docs/source/config/settings.rst
Settings
========

``indico.conf`` is Indico's main configuration file. Its initial version
is usually generated when running ``indico setup wizard`` as described in
the Installation Guide, but depending on the setup it should be modified
later.

The config file is loaded from the path specified in the ``INDICO_CONFIG``
environment variable; if no such path is set, the config file (or a symlink
to it) is searched in the following places, in order:

- ``<indico_package_path>/indico.conf`` (development setups only)
- ``~/.indico.conf``
- ``/etc/indico.conf``

The file is executed as a Python module, so anything that is valid Python
2.7 code can be used in it. When defining temporary variables that are not
config options, their name should be prefixed with an underscore; otherwise
you will get a warning about unknowing config options being defined.


.. _settings-auth:

Authentication
--------------

.. data:: LOCAL_IDENTITIES

    This setting controls whether local Indico accounts are available.
    If no centralized authentication infrastructure (e.g. LDAP, OAuth,
    or another kind of SSO) is used, local accounts are the only way
    of logging in to Indico.

    Default: ``True``

.. data:: LOCAL_GROUPS

    This setting controls whether local Indico groups are available.
    If no centralized authentication infrastructure that supports groups
    (e.g. LDAP) is used, local groups are the only way to define groups in
    Indico, but if you do have central groups it may be useful to disable
    local ones to have all groups in one central place.

    Default: ``True``

.. data:: LOCAL_REGISTRATION

    This setting controls whether people accessing Indico can create a
    new account.  Admins can always create new local accounts, regardless
    of this setting.

    This setting is only taken into account if :data:`LOCAL_IDENTITIES`
    are enabled.

    Default: ``True``

.. data:: LOCAL_MODERATION

    This setting controls whether a new registration needs to be approved
    by an admin before the account is actually created.

    This setting is only taken into account if :data:`LOCAL_IDENTITIES`
    and :data:`LOCAL_REGISTRATION` are enabled.

    Default: ``False``

.. data:: FAILED_LOGIN_RATE_LIMIT

    Applies a rate limit to failed login attempts due to an invalid username
    or password. When specifying multiple rate limits separated with a semicolon,
    they are checked in that specific order, which can allow for a short burst of
    attempts (e.g. a legitimate user trying multiple passwords they commonly use)
    and then slowing down more strongly (in case someone tries to brute-force more
    than just a few passwords).

    Rate limiting is applied by IP address and only failed logins count against the
    rate limit. It also does not apply to login attempts using external login systems
    (SSO) as failures there are rarely related to invalid credentials coming from the
    user (these would be rejected on the SSO side, which should implement its own rate
    limiting).

    The default allows a burst of 10 attempts, and then only 5 attempts every 15
    minutes for the next 24 hours.  Setting the rate limit to ``None`` disables it.

    Default: ``'5 per 15 minutes; 10 per day'``

.. data:: SIGNUP_RATE_LIMIT

    Applies a rate limit to sending verification emails in signup attempts.
    When specifying multiple rate limits separated with a semicolon, they are checked
    in that specific order, which can allow for a short burst of attempts.

    Rate limiting is applied by IP address and each verification email sent counts
    against the rate limit.

    The default allows a burst of 5 attempts, and then only 2 attempts every hour for
    the next 24 hours.  Setting the rate limit to ``None`` disables it.

    Default: ``'2 per hour; 5 per day'``

.. data:: EXTERNAL_REGISTRATION_URL

    The URL to an external page where people can register an account that
    can then be used to login to Indico (usually via LDAP/SSO).

    This setting is only taken into account if :data:`LOCAL_IDENTITIES`
    are disabled.

    Default: ``None``

.. data:: AUTH_PROVIDERS

    A dict defining `Flask-Multipass`_ authentication providers used
    by Indico.  The dict specified here is passed to the
    ``MULTIPASS_AUTH_PROVIDERS`` setting of Flask-Multipass.

    Default: ``{}``

.. data:: IDENTITY_PROVIDERS

    A dict defining `Flask-Multipass`_ identity providers used by Indico
    to look up user information based on the data provided by an
    authentication provider.  The dict specified here is passed to the
    ``MULTIPASS_IDENTITY_PROVIDERS`` setting of Flask-Multipass.

    Default: ``{}``

.. data:: PROVIDER_MAP

    If not specified, authentication and identity providers with the
    same name are linked automatically.  The dict specified here is
    passed to the ``MULTIPASS_PROVIDER_MAP`` setting of Flask-Multipass.

    Default: ``{}``

.. data:: SIGNUP_CAPTCHA

    If enabled, a CAPTCHA is required when creating an Indico account to prevent
    spam bots from registering accounts automatically.  Signups through external
    authentication systems (LDAP, SSO etc.) are not affected by this; they are
    expected to have their own protection in place to prevent spam signups.

    Default: ``True``


Cache
-----

.. data:: REDIS_CACHE_URL

    The URL of the redis server to use for caching.

    If the Redis server requires authentication, use a URL like this:
    ``redis://unused:password@127.0.0.1:6379/1``

    If no authentication is used (usually the case with a local Redis
    server), you can omit the user/password part:
    ``redis://127.0.0.1:6379/1``

    Default: ``None``

.. data:: MEMCACHED_SERVERS

    The list of memcached servers (each entry is an ``ip:port`` string)
    to use with the ``memcached`` cache backend.

    Default: ``[]``


Celery
------

.. data:: CELERY_BROKER

    The URL of the Celery broker (usually Redis of AMQP) used for
    communication between Indico and the Celery background workers.

    We recommend using Redis as it is the easiest option, but you can
    check the `Celery documentation on brokers`_ for more information
    on the other possible brokers.

    Default: ``None``

.. data:: CELERY_RESULT_BACKEND

    The URL of the Celery result backend. If not set, the same backend
    as the broker is used.  Indico currently does not use task results,
    and we recommend leaving this setting at its default.

    Default: ``None``

.. data:: CELERY_CONFIG

    A dict containing additional Celery settings.

    .. warning::

        This is an advanced setting that is rarely needed and we do not
        recommend using it unless you know exactly what you are doing!
        Changing Celery settings may break things or result in tasks not
        being executed without other changes (such as running additional
        celery workers on different queues).

    One use case for this setting is routing certain tasks to a different
    queue, and then running multiple Celery workers for these queues.

    .. code-block:: python

        CELERY_CONFIG = {
            'task_routes': {
                'indico_livesync.task.scheduled_update': {'queue': 'livesync'},
            }
        }

    Default: ``{}``

.. data:: SCHEDULED_TASK_OVERRIDE

    A dict overriding the task schedule for specific tasks.

    By default, all periodic tasks are enabled and use a schedule which
    we consider useful for most cases.  Using this setting, you can
    override the default schedule.

    The dict key is the name of the task and the value can be one of
    the following:

    - ``None`` or ``False`` -- disables the task completely
    - A dictionary, as described in the `Celery documentation on periodic tasks`_.
      The ``task`` should not be specified, as it is set automatically.
    - A :class:`~datetime.timedelta` or :class:`~celery.schedules.crontab`
      object which will just override the schedule without changing any
      other options of the task.  Both classes are available in the config
      file by default.

    .. note::

        Use ``indico celery inspect registered`` to get a list of task
        names.  Celery must be running for this command to work.

    Default: ``{}``


Customization
-------------

.. data:: CUSTOMIZATION_DIR

    The base path to the directory containing customizations for your
    Indico instance.

    It is possible to override specific templates and add CSS and
    JavaScript for advanced customizations.  When using this, be
    advised that depending on the modifications you perform things
    may break after an Indico update.  Make sure to test all your
    modifications whenever you update Indico!

    To include custom CSS and JavaScript, simply put ``*.css`` and
    ``*.js`` files into ``<CUSTOMIZATION_DIR>/css`` / ``<CUSTOMIZATION_DIR>/js``.
    If there are multiple files, they will be included in alphabetical
    order, so prefixing them with a number (e.g. ``00-base.css``, ``10-events.css``)
    is a good idea.

    Static files may be added in ``<CUSTOMIZATION_DIR>/files``.  They can be
    referenced in templates through the ``assets.custom`` endpoint.  In CSS/JS,
    the URL for them needs to be built manually (``/static/custom/files/...``).

    For template customizations, see the description of :data:`CUSTOMIZATION_DEBUG`
    as this setting is highly recommended to figure out where exactly to
    put customized templates.

    Here is an example for a template customization that includes a
    custom asset and uses inheritance to avoid having to replace the
    whole template:

    .. code-block:: jinja

        {% extends '~footer.html' %}

        {% block footer_logo %}
            {%- set filename = 'cern_small_light.png' if dark|default(false) else 'cern_small.png' -%}
            <a href="https://home.cern/" class="footer-logo">
                <img src="{{ url_for('assets.custom', filename=filename) }}" alt="CERN">
            </a>
        {% endblock %}

    Default: ``None``

.. data:: CUSTOMIZATION_DEBUG

    Whether to log details for all customizable templates the first time
    they are accessed.  The log message contains the path where you need
    to store the template; this path is relative to
    ``<CUSTOMIZATION_DIR>/templates/``.

    The log message also contains the full path of the original template
    in case you decide to copy it.
    However, instead of copying templates it is better to use Jinja
    inheritance where possible.  To make this easier the log entry contains
    a "reference" path that can be used to reference the original template
    from the customized one.

    Default: ``False``

.. data:: HELP_URL

    The URL used for the "Help" link in the footer.

    Default: ``'https://learn.getindico.io'``

.. data:: FAVICON_URL

    The URL to a custom favicon.  If unset, the default monochrome Indico favicon is
    used.

    Default: ``None``

.. data:: LOGO_URL

    The URL to a custom logo.  If unset, the default monochrome Indico logo is
    used.

    Default: ``None``

.. data:: LOGIN_LOGO_URL

    The URL to a custom logo used on the login page.  If unset, the default
    Indico logo is used.

    Default: ``None``

.. data:: WALLET_LOGO_URL

    The URL to a custom logo used in Google Wallet and Apple Wallet tickets.
    If unset, a compact version of the default Indico logo is used.
    The URL must publicly accessible from the Internet and changes typically
    do not affect existing tickets.

    Default: ``None``

.. data:: CUSTOM_COUNTRIES

    A dict with country name overrides.  This can be useful if the official
    ISO name of a country does not match what your Indico instance's target
    audience expects for a country, e.g. due to political situations.

    .. code-block:: python

        CUSTOM_COUNTRIES = {'KP': 'North Korea'}

    Default: ``{}``

.. data:: CUSTOM_LANGUAGES

    A dict with language/territory name overrides.  This can be useful if the
    official territory name that goes along with a language does not match what
    your Indico instance's target audience expects for a country, e.g. due to
    political situations.

    For example, to replace "Chinese (Simplified)" with "Chinese (China)",
    you would use:

    .. code-block:: python

        CUSTOM_LANGUAGES = {'zh_Hans_CN': ('Chinese', 'Simplified')}

    Note that the language and territory name should be written in that
    particular language to be consistent with the defaults. So in the example
    above, you would write "Chinese" and "Simplified" in Simplified Chinese.

    Setting the territory (second element in the tuple) to ``None`` will hide
    it and only show the language name itself.  Setting the dict value to ``None``
    will effectively hide the language altogether.

    Default: ``{}``

.. data:: CHECKIN_APP_URL

    The URL of the mobile checkin app. The app is purely client-side and only
    communicates with your Indico instance, so even when using the default app
    (which is hosted in CERN's datacenter in Switzerland) no data about your
    events or participants is sent to CERN, the Indico dev team or anyone else.
    If you wish to use a custom app nonetheless, you can find its
    `source code on GitHub <https://github.com/indico/indico-checkin-pwa/>`_ and
    deploy it wherever you want. Note that you need to add the URL of the app
    to the *"Allowed authorization callback URLs"* of the OAuth app named
    "Checkin App" in the Indico admin area.

    Default: ``'https://checkin.getindico.io'``


Database
--------

.. data:: SQLALCHEMY_DATABASE_URI

    The URI used to connect to the PostgreSQL database.  For a local database,
    you can usually omit everything besides the database name:
    ``postgresql:///indico``

    If the database requires authentication and/or runs on a separate host,
    this form should be used: ``postgresql://user:password@hostname/dbname``

.. data:: SQLALCHEMY_POOL_SIZE

    This setting configures SQLAlchemy's connection pool.
    For details, check the `SQLAlchemy connection pool documentation`_.

    Default: ``5``

.. data:: SQLALCHEMY_POOL_RECYCLE

    This setting configures SQLAlchemy's connection pool.
    For details, check the `SQLAlchemy connection pool documentation`_.

    Default: ``120``

.. data:: SQLALCHEMY_POOL_TIMEOUT

    This setting configures SQLAlchemy's connection pool.
    For details, check the `SQLAlchemy connection pool documentation`_.

    Default: ``10``


Development
-----------

.. warning::

    Do not turn on development settings in production.  While we are not
    aware of serious security issues caused by these settings, they may
    slow down Indico or remove redundancies and thus make Indico not as
    stable as one would expect it to be in a production environment.

.. data:: DEBUG

    Enables debugging mode.  If enabled, assets are not minified, error
    messages are more verbose and various other features are configured
    in a developer-friendly way.

    **Do not enable debug mode in production.**

    Default: ``False``

.. data:: DB_LOG

    Enables real-time database query logging.  When enabled, all database
    queries are sent to a socket where they can be read by the ``db_log.py``
    script.  To use the database logger, run ``bin/utils/db_log.py`` (only
    available when running Indico from a Git clone) in a separate terminal
    and all requests and verbose queries will be displayed there.

    Default: ``False``

.. data:: PROFILE

    Enables the Python profiler.  The profiler output is stored in
    ``<TEMP_DIR>/*.prof``.

    Default: ``False``

.. data:: SMTP_USE_CELERY

    If disabled, emails will be sent immediately instead of being
    handed to a Celery background worker.  This is often more convenient
    during development as you do not need to run a Celery worker while still
    receiving emails sent from Indico.
    Disabling it may result in emails not being sent if the mail server is
    unavailable or some other failure happens during email sending.  Because
    of this, the setting should never be disabled in a production environment.

    Default: ``True``

.. data:: COMMUNITY_HUB_URL

    The URL of the community hub. This should only be changed when using a local
    instance of Mereswine to debug the interface between Indico and Mereswine.

    Default: ``'https://hub.getindico.io'``

.. data:: SYSTEM_NOTICES_URL

    The URL of a YAML file with system notices. This should only be changed during
    development (to test custom notices) or set to ``None`` to opt-out from ever
    fetching or displaying system notices.

    Default: ``'https://getindico.io/notices.yml'``

.. data:: DISABLE_CELERY_CHECK

    Disables the warning about Celery not running or being outdated. When set to
    ``None``, the warning is disabled when :data:`DEBUG` is enabled; otherwise
    this setting enables/disables the warning regardless of debug mode.

    Default: ``None``


Directories
-----------

.. data:: CACHE_DIR

    The directory in which various data is cached temporarily. Must be
    accessible by the web server.

    Default: ``'/opt/indico/cache'``

.. data:: LOG_DIR

    The directory in which log files are stored. Can be overridden by
    using absolute paths in ``logging.yaml``.

    Default: ``'/opt/indico/log'``

.. data:: TEMP_DIR

    The directory in which various temporary files are stored. Must be
    accessible by the web server.

    Default: ``'/opt/indico/cache'``


Emails
------

.. data:: EMAIL_BACKEND

    Qualified import name for the email sending backend. It can be set to any email
    backend compatible with Django Mail.

    Default: ``'indico.vendor.django_mail.backends.smtp.EmailBackend'``

.. data:: SMTP_SERVER

    The hostname and port of the SMTP server used for sending emails.

    Default: ``('localhost', 25)``

.. data:: SMTP_LOGIN

    The username to send if the SMTP server requires authentication.

    Default: ``None``

.. data:: SMTP_PASSWORD

    The password to send if the SMTP server requires authentication.

    Default: ``None``

.. data:: SMTP_USE_TLS

    If enabled, STARTTLS will be used to use an encrypted SMTP connection.

    Default: ``False``

.. data:: SMTP_CERTFILE

    If provided, this certificate file will be used for certificate-based
    SMTP authentication.

    Default: ``None``

.. data:: SMTP_KEYFILE

    If provided, this private key file will be used for certificate-based
    SMTP authentication.

    Default: ``None``

.. data:: SMTP_TIMEOUT

    The timeout in seconds after which a connection attempt to the SMTP
    server is aborted.

    Default: ``30``

.. data:: SMTP_ALLOWED_SENDERS

    A list of allowed envelope sender addresses. Each entry must be an email
    address, but using the ``*`` wildcard is allowed.

CHANGES.rst
Changelog
=========


Version 3.3.4
-------------

*Unreleased*

Improvements
^^^^^^^^^^^^

- Allow cropping an existing picture in registration form picture fields (:pr:`6423`,
  thanks :user:`SegiNyn`)
- Add task to delete old registration files when they become orphaned due to a new
  file being uploaded (:pr:`6434`, thanks :user:`SegiNyn`)

Bugfixes
^^^^^^^^

- Fix display of empty session selection in registration summary (:pr:`6421`,
  thanks :user:`jbtwist`)
- Include date when displaying session field data in registration summary (:pr:`6431`,
  thanks :user:`jbtwist`)
- Fix the order of a day's session blocks in the registration form session field
  (:pr:`6428`, thanks :user:`jbtwist`)
- Wrap overly long descriptions and filenames in registration form fields (:pr:`6436`,
  thanks :user:`SegiNyn`)

Internal Changes
^^^^^^^^^^^^^^^^

- Indicate when a booking begins/ends in the booking calendar in day-based mode (when
  using a plugin to customize the room booking module) (:pr:`6414`)
- Update the list of supported browsers so people using highly outdated browsers where
  certain features are likely broken get a warning about having to update their browser
  (:pr:`6442`)


Version 3.3.3
-------------

*Released on June 26, 2024*

Internationalization
^^^^^^^^^^^^^^^^^^^^

- New translation: Hungarian

Improvements
^^^^^^^^^^^^

- Add dialog to contact event participants about a survey (:issue:`6069`, :pr:`6144`)
- Allow linking existing room booking occurrences to an event (:pr:`6243`, thanks
  :user:`Moliholy, unconventionaldotdev`)
- Support including a picture (from a registration's picture field) in the conference
  participant list (:pr:`6228`, thanks :user:`vtran99`)
- Add :data:`FAVICON_URL` config option to set a custom URL for the favicon (:pr:`6323`,
  thanks :user:`SegiNyn`)
- Allow filtering the contribution list in the management area by custom fields
  (:issue:`6213`, :pr:`6214`)
- Show "Go to timeline" button on the contribution page to everyone who can see the
  timeline of one of its editables instead of just submitters (:pr:`6344`)
- Add a new "Timetable Sessions" registration form field type which allows selecting
  session blocks from the event (:pr:`6184`, thanks :user:`jbtwist`)
- Link the event title to the event in registration emails (:pr:`6358`)
- Add the option to make registration forms private so they can only be accessed using
  a secret link (:pr:`6321`, thanks :user:`vtran99`)
- Add experimental support for creating Apple Wallet (Passbook / pkpass) tickets
  (opt-in via :data:`ENABLE_APPLE_WALLET` ``indico.conf`` setting) (:pr:`6248`, thanks
  :user:`openprojects`)
- Add a new event management permission that grants access only to the contributions
  module (:pr:`6348`)
- Add bulk JSON export option in management contribution list (:pr:`6370`)
- Make the default roles of the contribution person link list field more similar to the
  abstract person link list field when there is a linked abstract (:pr:`6342`)
- Add option to hide person titles throughout the event (:issue:`038`, :pr:`6104`, thanks
  :user:`vasantvohra`)
- Preserve input when switching between judgment actions for an editable (:pr:`6375`)
- Allow generating documents from the registration summary page (:issue:`6212`, :pr:`6306`,
  thanks :user:`hitenvidhani`)
- Modernize the event social share widget and add support for sharing to
  Mastodon (:pr:`6289`)
- Enable the calendaring + social sharing widget in events by default (:pr:`6398`)
- Ignore diacritics when searching in the registration form country field (:pr:`6403`,
  thanks :user:`tomako`)
- Add preview option for managers to see the participant list as shown to registered
  participants or unregistered guests (:pr:`6052`, thanks :user:`vtran99`)

Bugfixes
^^^^^^^^

- Fix the dashboard iCal export returning old events instead of recent ones when the
  maximum number of events to include is reached (:pr:`6312`)
- Fix an error in the Check-in app API wben retrieving details for a registration form
  that includes static labels (:pr:`6326`)
- Fix action buttons being pushed outside the content area in the survey editor in case
  of very long survey option titles (:pr:`6325`)
- Only allow accessing avatars for published registrations (:pr:`6347`)
- Fix error when trying to import data from an unlisted event (:issue:`6350`, :pr:`6351`)
- Show results from the Get Next Editable search on top of the list (:pr:`6353`)
- Attach registration pictures and display them inline when sending email notifications
  instead of just showing their filename (:pr:`6336, 6411`, thanks :user:`SegiNyn`)
- Fix editable list filter storage being shared between different editable types and
  events (:pr:`6359`)
- Fix UI breaking when performing bulk actions via the list of editables (:pr:`6369`)
- Include registration documents in user data export (:issue:`6331`, :pr:`6338`)
- Fix error when viewing an abstract with reviews in deleted tracks (:pr:`6393`)
- Do not include custom messages about the current registration status when sending
  notifications about new documents (:pr:`6413`)
- Only normalize title slug in custom page URL after successful access check
  (:issue:`6416`, :pr:`6417`)

Accessibility
^^^^^^^^^^^^^

- Improve registration form date picker accessibility (:pr:`6371`, thanks :user:`foxbunny`)

Internal Changes
^^^^^^^^^^^^^^^^

- Use unguessable URLs for user avatar pictures (:pr:`6346`, thanks :user:`vtran99`)
- Add ``<ind-date-picker>`` custom element (:pr:`6371, 6406`, thanks :user:`foxbunny`)
- Use native ESM for webpack config files (:pr:`6389`)
- Rename ``active_fields`` to ``available_fields`` in ``RegistrationFormSection``
  (:pr:`6409`, thanks :user:`omegak`)
- Custom poster/badge designer placeholder returning images need to return a ``BytesIO``
  instead of a Pillow ``Image`` object (:pr:`6441`)


Version 3.3.2
-------------

*Released on April 19, 2024*

Improvements
^^^^^^^^^^^^


indico/core/config.py
# This file is part of Indico.
# Copyright (C) 2002 - 2024 CERN
#
# Indico is free software; you can redistribute it and/or
# modify it under the terms of the MIT License; see the
# LICENSE file for more details.

import ast
import codecs
import os
import re
import socket
import warnings
from datetime import timedelta
from urllib.parse import urljoin, urlsplit

import pytz
from celery.schedules import crontab
from flask import current_app, g
from flask.helpers import get_root_path
from werkzeug.datastructures import ImmutableDict

from indico.util.caching import make_hashable
from indico.util.fs import resolve_link
from indico.util.packaging import package_is_editable
from indico.util.string import crc32, snakify


# Note: Whenever you add/change something here, you MUST update the docs (settings.rst) as well
DEFAULTS = {
    'ALLOW_PUBLIC_USER_SEARCH': True,
    'ATTACHMENT_STORAGE': 'default',
    'AUTH_PROVIDERS': {},
    'BASE_URL': None,
    'CACHE_DIR': '/opt/indico/cache',
    'CATEGORY_CLEANUP': {},
    'CELERY_BROKER': None,
    'CELERY_CONFIG': {},
    'CELERY_RESULT_BACKEND': None,
    'CHECKIN_APP_URL': 'https://checkin.getindico.io',
    'COMMUNITY_HUB_URL': 'https://hub.getindico.io',
    'CUSTOMIZATION_DEBUG': False,
    'CUSTOMIZATION_DIR': None,
    'CUSTOM_COUNTRIES': {},
    'CUSTOM_LANGUAGES': {},
    'DB_LOG': False,
    'DEBUG': False,
    'DEFAULT_LOCALE': 'en_GB',
    'DEFAULT_TIMEZONE': 'UTC',
    'DISABLE_CELERY_CHECK': None,
    'EMAIL_BACKEND': 'indico.vendor.django_mail.backends.smtp.EmailBackend',
    'ENABLE_APPLE_WALLET': False,
    'ENABLE_GOOGLE_WALLET': False,
    'ENABLE_ROOMBOOKING': False,
    'EXPERIMENTAL_EDITING_SERVICE': False,
    'EXTERNAL_REGISTRATION_URL': None,
    'HELP_URL': 'https://learn.getindico.io',
    'FAILED_LOGIN_RATE_LIMIT': '5 per 15 minutes; 10 per day',
    'FAVICON_URL': None,
    'IDENTITY_PROVIDERS': {},
    'LOCAL_IDENTITIES': True,
    'LOCAL_MODERATION': False,
    'LOCAL_REGISTRATION': True,
    'LOCAL_GROUPS': True,
    'LOGGING_CONFIG_FILE': 'logging.yaml',
    'LOGIN_LOGO_URL': None,
    'LOGO_URL': None,
    'LOG_DIR': '/opt/indico/log',
    'MAX_DATA_EXPORT_SIZE': 10 * 1024,  # 10GB
    'MAX_UPLOAD_FILES_TOTAL_SIZE': 0,
    'MAX_UPLOAD_FILE_SIZE': 0,
    'MEMCACHED_SERVERS': [],
    'NO_REPLY_EMAIL': None,
    'PLUGINS': set(),
    'PROFILE': False,
    'PROVIDER_MAP': {},
    'PUBLIC_SUPPORT_EMAIL': None,
    'REDIS_CACHE_URL': None,
    'ROUTE_OLD_URLS': False,
    'SCHEDULED_TASK_OVERRIDE': {},
    'SECRET_KEY': None,
    'SENTRY_DSN': None,
    'SENTRY_LOGGING_LEVEL': 'WARNING',
    'SESSION_LIFETIME': 86400 * 31,
    'SIGNUP_CAPTCHA': True,
    'SIGNUP_RATE_LIMIT': '2 per hour; 5 per day',
    'SMTP_ALLOWED_SENDERS': set(),
    'SMTP_CERTFILE': None,
    'SMTP_KEYFILE': None,
    'SMTP_LOGIN': None,
    'SMTP_PASSWORD': None,
    'SMTP_SENDER_FALLBACK': None,
    'SMTP_SERVER': ('localhost', 25),
    'SMTP_TIMEOUT': 30,
    'SMTP_USE_CELERY': True,
    'SMTP_USE_TLS': False,
    'SQLALCHEMY_DATABASE_URI': None,
    'SQLALCHEMY_MAX_OVERFLOW': 3,
    'SQLALCHEMY_POOL_RECYCLE': 120,
    'SQLALCHEMY_POOL_SIZE': 5,
    'SQLALCHEMY_POOL_TIMEOUT': 10,
    'STATIC_FILE_METHOD': None,
    'STATIC_SITE_STORAGE': None,
    'STORAGE_BACKENDS': {'default': 'fs:/opt/indico/archive'},
    'STRICT_LATEX': False,
    'SUPPORT_EMAIL': None,
    'SYSTEM_NOTICES_URL': 'https://getindico.io/notices.yml',
    'TEMP_DIR': '/opt/indico/tmp',
    'USE_PROXY': False,
    'WALLET_LOGO_URL': None,
    'WORKER_NAME': socket.getfqdn(),
    'XELATEX_PATH': None,
}

# Default values for settings that cannot be set in the config file
INTERNAL_DEFAULTS = {
    'CONFIG_PATH': os.devnull,
    'CONFIG_PATH_RESOLVED': None,
    'LOGGING_CONFIG_PATH': None,
    'TESTING': False
}


def get_config_path():
    """Get the path of the indico config file.

    This may return the location of a symlink.  Resolving a link is up
    to the caller if needed.
    """
    # In certain environments (debian+uwsgi+no-systemd) Indico may run
    # with an incorrect $HOME (such as /root), resulting in the config
    # files being searched in the wrong place. By clearing $HOME, Python
    # will get the home dir from passwd which has the correct path.
    old_home = os.environ.pop('HOME', None)
    # env var has priority
    try:
        return os.path.expanduser(os.environ['INDICO_CONFIG'])
    except KeyError:
        pass
    # try finding the config in various common paths
    paths = [os.path.expanduser('~/.indico.conf'), '/etc/indico.conf']
    # Keeping HOME unset wouldn't be too bad but let's not have weird side-effects
    if old_home is not None:
        os.environ['HOME'] = old_home
    # If it's an editable setup (ie usually a dev instance) allow having
    # the config in the package's root path
    if package_is_editable('indico'):
        paths.insert(0, os.path.normpath(os.path.join(get_root_path('indico'), 'indico.conf')))
    for path in paths:
        if os.path.exists(path):
            return path
    raise Exception('No indico config found. Point the INDICO_CONFIG env var to your config file or '
                    'move/symlink the config in one of the following locations: {}'.format(', '.join(paths)))


def _parse_config(path):
    globals_ = {'timedelta': timedelta, 'crontab': crontab}
    locals_ = {}
    with codecs.open(path, encoding='utf-8') as config_file:
        exec(compile(config_file.read(), path, 'exec'), globals_, locals_)  # noqa: S102
    return {str(k if k.isupper() else _convert_key(k)): v
            for k, v in locals_.items()
            if k[0] != '_'}


def _convert_key(name):
    # camelCase to BIG_SNAKE while preserving acronyms, i.e.
    # FooBARs -> FOO_BARS (and not FOO_BA_RS)
    name = re.sub(r'([A-Z])([A-Z]+)', lambda m: m.group(1) + m.group(2).lower(), name)
    name = snakify(name).upper()
    special_cases = {'PDFLATEX_PROGRAM': 'XELATEX_PATH',
                     'IS_ROOM_BOOKING_ACTIVE': 'ENABLE_ROOMBOOKING'}
    return special_cases.get(name, name)


def _postprocess_config(data):
    data['BASE_URL'] = data['BASE_URL'].rstrip('/')
    data['STATIC_SITE_STORAGE'] = data['STATIC_SITE_STORAGE'] or data['ATTACHMENT_STORAGE']
    if data['DISABLE_CELERY_CHECK'] is None:
        data['DISABLE_CELERY_CHECK'] = data['DEBUG']


def _sanitize_data(data, allow_internal=False):
    allowed = set(DEFAULTS)
    if allow_internal:
        allowed |= set(INTERNAL_DEFAULTS)
    for key in set(data) - allowed:
        warnings.warn(f'Ignoring unknown config key {key}', stacklevel=2)
    return {k: v for k, v in data.items() if k in allowed}


def load_config(only_defaults=False, override=None):
    """Load the configuration data.

    :param only_defaults: Whether to load only the default options,
                          ignoring any user-specified config file
                          or environment-based overrides.
    :param override: An optional dict with extra values to add to
                     the configuration.  Any values provided here
                     will override values from the config file.
    """
    data = DEFAULTS | INTERNAL_DEFAULTS
    if not only_defaults:
        path = get_config_path()
        config = _sanitize_data(_parse_config(path))
        data.update(config)
        env_override = os.environ.get('INDICO_CONF_OVERRIDE')
        if env_override:
            data.update(_sanitize_data(ast.literal_eval(env_override)))
        resolved_path = resolve_link(path) if os.path.islink(path) else path
        resolved_path = None if resolved_path == os.devnull else resolved_path
        data['CONFIG_PATH'] = path
        data['CONFIG_PATH_RESOLVED'] = resolved_path
        if resolved_path is not None:
            data['LOGGING_CONFIG_PATH'] = os.path.join(os.path.dirname(resolved_path), data['LOGGING_CONFIG_FILE'])

    if override:
        data.update(_sanitize_data(override, allow_internal=True))
    _postprocess_config(data)
    return ImmutableDict(data)


class IndicoConfig:
    """Wrapper for the Indico configuration.

    It exposes all config keys as read-only attributes.

    Dynamic configuration attributes whose value may change depending
    on other factors may be added as properties, but this should be
    kept to a minimum and is mostly there for legacy reasons.

    :param config: The dict containing the configuration data.
                   If omitted, it is taken from the active flask
                   application.  An explicit configuration dict should
                   not be specified except in special cases such as
                   the initial app configuration where no app context
                   is available yet.
    :param exc: The exception to raise when accessing an invalid
                config key.  This allows using the expected kind of
                exception in most cases but overriding it when
                exposing settings to Jinja where the default
                :exc:`AttributeError` would silently be turned into
                an empty string.
    """

    __slots__ = ('_config', '_exc')

    def __init__(self, config=None, exc=AttributeError):
        # yuck, but we don't allow writing to attributes directly
        object.__setattr__(self, '_config', config)
        object.__setattr__(self, '_exc', exc)

    @property
    def data(self):
        try:
            return self._config or current_app.config['INDICO']
        except KeyError:
            raise RuntimeError('config not loaded')

    @property
    def hash(self):
        return crc32(repr(make_hashable(sorted(self.data.items()))))

    @property
    def CONFERENCE_CSS_TEMPLATES_BASE_URL(self):
        return self.BASE_URL + '/css/confTemplates'

    @property
    def IMAGES_BASE_URL(self):
        return 'static/images' if g.get('static_site') else urlsplit(f'{self.BASE_URL}/images').path

    @property
    def LATEX_ENABLED(self):
        return bool(self.XELATEX_PATH)

    @property
    def ABSOLUTE_LOGO_URL(self):
        return urljoin(self.BASE_URL, self.LOGO_URL) if self.LOGO_URL else None

    @property
    def ABSOLUTE_WALLET_LOGO_URL(self):
        return urljoin(self.BASE_URL, self.WALLET_LOGO_URL) if self.WALLET_LOGO_URL else None

    def validate(self):
        from indico.core.auth import login_rate_limiter
        login_rate_limiter._get_current_object()  # fail in case FAILED_LOGIN_RATE_LIMIT invalid
        if self.DEFAULT_TIMEZONE not in pytz.all_timezones_set:
            raise ValueError(f'Invalid default timezone: {self.DEFAULT_TIMEZONE}')
        if self.SMTP_ALLOWED_SENDERS and not self.SMTP_SENDER_FALLBACK:
            raise ValueError('Cannot restrict SMTP senders without a fallback')

    def __getattr__(self, name):
        try:
            return self.data[name]
        except KeyError:
            raise self._exc('no such setting: ' + name)

    def __setattr__(self, key, value):
        raise AttributeError('cannot change config at runtime')

    def __delattr__(self, key):
        raise AttributeError('cannot change config at runtime')


#: The global Indico configuration
config = IndicoConfig()
